### DISCUSSION TITLE
Native pipewire support on Linux

### DISCUSSION THREAD

**Discussion by f1ori** [jamulussoftware/jamulus#1454] [Ideas]
Hi,
is there native pipewire support on Linux planed or anybody working on this? With Fedora 34, the first Linux distribution makes pipewire the default. Pipewire would make configuration much easier, because it handles most audio use cases and doesn't need a special setup like jack.

**Comment by hoffie** [comment]
I think I remember that the Flatpak builds required pipewire support and it seemed to work, but I haven't tried myself yet. Maybe you can find the relevant code, scripts or configs by looking in the issues/discussions and/or on Flathub (?). I also assume that this made use of some compatibility (?) mode from pipewire.

So, without knowing any details about pipewire, but assuming that it is a new set of APIs, I'm pretty sure that there is no native support yet and I haven't seen any open issues or PRs about it either.

Have you tried using Jamulus with pipewire?

**Reply by f1ori** [reply]
I use the flatpak and the pipewire compatibility mode for jack works fine.
What was inconvenient is, that I needed to use qjackctl to assign the correct audio devices and I need to open jamulus with the environment variable PIPEWIRE_LATENCY="128/48000", to configure the latency.
If nobody else is working on this, I will try to add native pipewire support, though I can't promise anything.

**Reply by hoffie** [reply]
I think anything which helps to make Jamulus work out of the box would be good.

I can't judge what exactly native pipewire supports would mean, but I would assume that something similar to the current jack implementation would be needed. Maybe someone else from @jamulussoftware/maindevelopers can comment on whether this would be acceptable.
Personally I think it is.

**Reply by jujudusud** [reply]
@f1ori did you go any further?

**Reply by jujudusud** [reply]
> I think anything which helps to make Jamulus work out of the box would be good.
> 
> I can't judge what exactly native pipewire supports would mean, but I would assume that something similar to the current jack implementation would be needed. Maybe someone else from @jamulussoftware/maindevelopers can comment on whether this would be acceptable. Personally I think it is.

@hoffie , one thing to make Jamulus fully usable with PipeWire would be the implementation of runtime buffer size change.
There is a todo in the code about this. I actually read documentations to understand how to add this fonctionnality to Jamulus. Don't know for now how to do this in C++.

**Reply by jujudusud** [reply]
@jamulussoftware/maindevelopers Anything to say about adding runtime buffer size change on Linux?

**Reply by ann0see** [reply]
If you succeed go for it. But the audio functionality is quite hard to get right from my experience.

**Comment by jujudusud** [comment]
I think that pipewire comes to replace pulse audio which lives with Alsa by relying on the apulse library. Jack remains usable even in the presence of pipewire. As long as pipewire will not be at the level of Jack in terms of low audio latency and synchronization of the streams (jack transport) I think that it is not the place to go. 
On the other hand, even if it is very convenient and powerful, Qjackctl is not the only way to change JACK settings.
It is also possible to do this from the command line.

**Reply by Beiri22** [reply]
But isn't pipewire intended to replace pulse and jack?

**Reply by jujudusud** [reply]
Pipewire is made to replace JACK + PulseAudio + Gstreamer.

**Reply by jujudusud** [reply]
And recently -> ffado

**Comment by jujudusud** [comment]
For information :
I installed 'PipeWire' + 'pipewire-jack' + 'Jamulus'.
I added 'Patchance' to set the buffer size and to route audio.
It works fine.

**Comment by jujudusud** [comment]
I'm bringing this thread up again to talk about my experience with PipeWire and Jamulus and a linux machine.

First of all, with a few settings, it works.

I'm using archlinux as operating system, ALSA + PipeWire for the sound server. PipeWire is providing JACK with pipewire-jack.
PipeWire version: `0.3.77`
My kernel version: `Linux 6.4.11 SMP PREEMPT_DYNAMIC x86_64 GNU/Linux`

In order to have the best results in term of sound quality and lowest latency, I have to force the samplerate to 48 kHz and the buffer size to 128. I'm doing that with `PipeControl`.
If I don't apply these settings, I get an exploding delay and poorer sound quality.
Without settings applied: Ping 18 ms, Delay 95 ms, Jitter green. Some artifacts.
With settings applied: Ping 17 ms, Delay 53 ms, jitter green. Nearly no artifacts.

Just to let you know.

**Reply by pljones** [reply]
It might be worth writing this up as a KB entry.

**Reply by jujudusud** [reply]
Certainly.
What I'd like to know before writing this kind of "reference" document is whether there is a solution for launching software with the right samplerate and buffer size simply with a command line in a PipeWire setup instead of "bricking" the entire system.

**Reply by jujudusud** [reply]
Using :
`PIPEWIRE_LATENCY=128/48000 jamulus`
to lunch the application.
A good way to force Jamulus to use the right samplerate and buffer size.
There is another way, using `PipeWire's Jack rules` but, for now, I don't know how.

If anyone knows anything ... :-)

**Reply by jujudusud** [reply]
The best solution for now seems to be : https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Guide-JACK
```
pw-metadata -n settings 0 clock.force-rate 48000
pw-metadata -n settings 0 clock.force-quantum 128
```

**Comment by jujudusud** [comment]
Now that the PipeWire 1.0.0 is out and many Linux distros Comes with it as default, what is the plan to document this new use case?

**Reply by ann0see** [reply]
I don't have much experience with pipewire yet. But if it's becoming default, it should (must) be documented

**Reply by jujudusud** [reply]
Thanks ann0see.
I opened an issue :  number 979 of jamulus website [here](https://github.com/jamulussoftware/jamuluswebsite/pull/979)
I have to document that part. We should go to into this issue and the future PR to talk about the "how to document".
I have some ideas. I will explain as soon as possible what I'm thinking about that.

**Reply by jujudusud** [reply]
Done,

Please see PR of jamulus website number 980 : [here](https://github.com/jamulussoftware/jamuluswebsite/pull/980)
